internal/core/src/segcore/segment_c.cpp

```diff
@@ -150,6 +150,43 @@ Insert(CSegmentInterface c_segment,
     }
 }
 
+CStatus
+InsertColumnData(CSegmentInterface c_segment,
+                 int64_t reserved_offset,
+                 int64_t size,
+                 const int64_t* row_ids,
+                 const uint64_t* timestamps,
+                 void* raw_data,
```

Comment:
Is it possible to pass in an array of pointers. So raw_data doesn't need to be completely flattened?

Reply:
Columns of data from streams and batches both are multiple pointers. If you need to pass a flattened pointer, you have to copy the memory.
Am i right?

Reply:
The design generally follows the previous row-based insertion, which uses void* pointer to pass data flattened by go caller.

Reply:
I am unsure about whether the copy is redundant here. But as long as memory safety is guaranteed, we could remove copies in both insertion methods.

Reply:
> The design generally follows the previous row-based insertion, which uses void* pointer to pass data flattened by go caller.

The previous row-based insert is about to be deprecated. I think it's unnecessary to stick to the previous design.
If go caller wants to flatten multiple contiguous memories into one contiguous memory, it has to do a memory copy. Removing this copy operation on the go side can speed up and save peak memory usage.

Reply:
copy on the c++ side is inevitable.

Reply:
> Is it possible to pass in an array of pointers. So raw_data doesn't need to be completely flattened?

As suggested by the [cgo doc](https://pkg.go.dev/cmd/cgo), we cannot pass an array of go pointers. (Here go pointer means a pointer to memory allocated by Go)
> Go code may pass a Go pointer to C provided the Go memory to which it points does not contain any Go pointers.

Reply:
A possible solution would be something like `SegmentSealed::LoadFieldData` as mentioned by @xige-16, maybe named `SegmentGrowing::InsertFieldData`, which however requires a significant refactoring of the segcore module and is beyond the scope of this PR and my capacity as I don't quite know segcore.

Reply:
The original data will be marshaled according proto, and will no longer be spliced into bytes in the go code. Modifications will be in my next pr. @czs007